pipeline/04-interpret.R

```diff
@@ -105,6 +112,115 @@ lightgbm::lgb.importance(lgbm_final_full_fit$fit) %>%
   rename_with(~ paste0(.x, "_value"), gain:frequency) %>%
   write_parquet(paths$output$feature_importance$local)
 
+
+
+
+#- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+# 4. Calculate comps -----------------------------------------------------------
+#- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+
+if (comp_enable) {
+  message("Calculating comps")
+
+  # Calculate the leaf node assignments for every predicted value.
+  # Due to integer overflow problems with leaf node assignment, we need to
+  # chunk our data such that they are strictly less than the limit of 1073742
+  # rows. More detail here: https://github.com/microsoft/LightGBM/issues/1884
+  chunk_size <- 500000
+  chunks <- split(
+    assessment_data_prepped,
+    ceiling(seq_along(assessment_data_prepped[[1]]) / chunk_size)
+  )
+  chunked_leaf_nodes <- chunks %>%
+    map(\(chunk) {
+      predict(
+        object = lgbm_final_full_fit$fit,
+        newdata = as.matrix(chunk),
+        type = "leaf",
+      )
+    })
+  # Prefer do.call(rbind, ...) over bind_rows() because the chunks are
+  # not guaranteed to have the same number of rows, and bind_rows() will raise
+  # an error in that case
+  leaf_nodes <- do.call(rbind, chunked_leaf_nodes) %>% as_tibble()
+
+  # Calculate weights representing feature importance, so that we can weight
+  # leaf node assignments based on the most important features.
+  # To do this, we need the training data so that we can compute base model
+  # error
+  message("Extracting weights from training data")
+  training_data <- read_parquet(paths$input$training$local) %>%
+    filter(!ind_pin_is_multicard, !sv_is_outlier) %>%
+    as_tibble()
+
+  tree_weights <- extract_weights(
+    model = lgbm_final_full_fit$fit,
+    train = training_data,
+    outcome_col = "meta_sale_price",
+  )
+
+  # Get leaf node assignments for the training data. Assume that the training
+  # data is a subset of the assessment data
+  assessment_joined_to_training <- assessment_data %>%
+    tibble::rownames_to_column() %>%
+    inner_join(training_data, by = c("meta_pin", "meta_card_num"))
```

Comment:
I'm not actually sure if this is a fair assumption. Notably, the training data goes back to 2014, whereas the assessment set is all from the most recent year; if PINs change during that time, we may end up removing them from the comparison set. I'm not sure what the best way to handle this is, short of rerunning `predict()` on `training_data`.